### DISCUSSION TITLE
消去画像ファイルの扱い

### DISCUSSION THREAD

**Discussion by tsubolabo** [increments/qiita-discussions#344] [ヘルプ]
正規投稿予定の記事を、とりあえず限定共有記事としてアップし、推敲・編集の都度、その中の参照画像を納得いくまで何回も描き替えています。しかし、そのときに気になることがあります。

　編集エディタ画面上から、古い画像の<img ～>を消去し、新しい画像をドラッグ＆ドロップして差し替えたつもりになっていました。しかし、よくよく見れば、古い画像のURLが再利用されることはなく、その都度、新規URLが割り当てられているようです。消去した画像のQiita内URLを控えておいて、後日、ブラウザのURL欄に打ち込めば、消去したつもりの古い画像がそっくりそのまま表示されます。

　これに気づくと、私が画像を更新するたびに、今後利用予定もない古いゴミ画像ファイルによって、Qiitaのサーバーが埋め尽くされてしまうのでは？　という微かな背徳意識が湧いてきます。

　実際のところ、どうなのでしょう？　古いURLに新しい画像を上書き保存する方法はあるのでしょうか？　また、これらのゴミ画像も、月間の投稿画像容量制限に加算されるのでしょうか？

**Comment by kato647** [comment]
@tsubolabo 
現状、画像の上書き保存はできません。
また一度でもアップロードした画像は、月間の投稿画像容量制限に加算されます。
月に100MBまでであれば、自由に画像をアップロードしても大丈夫です！

**Reply by tsubolabo** [reply]
ご回答ありがとうございます。了解しました。個人的には、下記の理由により、現行システムのままでもあまり悩む必要はなさそうです。

+ Qiitaは、ゴミ画像の少々の累積はあまり気にされていなさそう。
+ 私の画像ファイルは、普通は１ファイル当たり数10KB程度なので、幾ら修正を繰り返しても、100MB/月を超えることはなさそう。

　なお、当ヘルプでの最初の質問は、画像だけでの消去や上書きを想定したものでした。しかし、記事全体を削除した場合には、その記事の添付画像（含む：ゴミ画像）は全部一緒に消去されると考えて良いのでしょうか？

**Comment by Phroneris** [comment]
横から失礼します。

私も詳しくはありませんけど…、例えばMastodonというオープンソースSNSでは、投稿に添付するためにアップロードされたメディアは投稿を消しても直ちに消去されるわけではなく、一旦「非添付」状態とでもいうべき様子見期間をおいてから1日後に消去されるらしい[^mastodon-media-delete]です。
書いたり消したりの利便性、画像の生URLを手入力せずに済む手軽さ、そして不要物の掃除とのバランスを取った結果、「この画像どこからも使われてないよ」と判断してくれる何らかの賢いシステムを間に挟むことによって良い感じにしているんでしょうね。

恐らくですが、QiitaとかGithub等のマークダウン系エディターや、もっと言えばTwitterとかDiscord等のより大衆向けでグラフィカルなエディターも、期間とかの設定の差異はあれど根幹部分の添付メディアの取り扱いは同じような仕様なんじゃないかと思います。
Twitterで消した画像のURLがいつまで生きてるかなんて自発的に気にしますか？（まぁ言われてみるとちょっと気にはなりますが…）
マークダウン系エディターだとなまじURLが見えるために気になってしまうだけではないでしょうか。

その控えておいた画像URLをもっと長く控えておいて、1ヶ月後とか半年後とかに見たら多分消えてると思われます。多分…

[^mastodon-media-delete]: https://github.com/mastodon/mastodon/issues/11637#issuecomment-523648546
（ただし2019年時点のお話です）

**Reply by tsubolabo** [reply]
[Phroneris](https://github.com/Phroneris) さん

　下記、拝見しました。大変勉強になりました。ありがとうございます。そのような運用が世の常識なのですね。少し待って消えるのであれば安心です。

> Media is not getting deleted after deleting post #11637


追伸：

> Twitterで消した画像のURLがいつまで生きてるかなんて自発的に気にしますか？

　SNSの類は、ここのQiita以外ではあまり経験がありませんので、Twitterのことは良く分かりません。しかし、宇宙デブリ、温室効果ガス、オゾン層破壊物質、海洋プラスチックなどの環境問題が話題になっていたせいか？　ゴミ画像ファイルのことを少々気にし過ぎていたのかもしれません。

> その控えておいた画像URLをもっと長く控えておいて、1ヶ月後とか半年後とかに見たら多分消えてると思われます。多分…

　私の下記の記述の中の「後日」は実は正確ではありません。適語が浮かばなかったので、ついつい適当に書いてそのままになっていました。正確には「後刻」です。ブラウザを全て閉じて（キャッシュやクッキーは強制消去の設定）、その数時間後に確認しただけです。１日待てば消えていたかもしれません。

> 後日、ブラウザのURL欄に打ち込めば、消去したつもりの古い画像がそっくりそのまま表示されます。

**Reply by tsubolabo** [reply]
下記の記事でも、Twitter での消去画像ファイルの扱いの実例が紹介されていました。


[最近気になってること - Qiita](https://qiita.com/m-m-n/items/69a930d80bc0f420a3a4)

> ちなみに画像を削除した場合は多少のタイムラグありますが、ちゃんと画像が失効するようでした

　この記事の本題は、画像の消去というよりも「限定者にしか見せないつもりの記事でも、その中の添付画像については、見ようとすれば誰にでも見られる」ということのようですが・・・。

　私も、限定共有記事なのに、その中の画像だけなら誰にでも見られるという点が気になっていました。

**Reply by Phroneris** [reply]
え、これ知らなかったです。
確かに技術的にはそういうもんなのかもしれませんが、直感には反する気がして驚きですね…！

**Reply by Phroneris** [reply]
@tsubolabo
すみません、私の元レスの「きっとQiitaのシステムが良い感じに不使用画像を削除してくれてるはず」的な推察は、恐らく間違いでした。

というのも最近、Qiitaに上げた画像を自力で削除できる機能が実装されました。
→ https://qiita.com/Aichi_Lover/items/b7f7224ae5bd378da5e3
この機能で私の「アップロードしたファイル」一覧を見てみたところ、過去（2018年とか）に記事編集の過程で一旦上げてから消したような明らかなボツ画像が、今でもその一覧に残っていました。
このことから、少なくとも不使用画像は全て自動で良い感じに消えてくれるわけではないようです。
必ず残るのか、残ったり残らなかったりするのかはわかりませんが…。

ということで、上げた画像の存在を完全に管理したいならこの機能で人力削除することになるでしょう。
ただ、ムーアの法則で例えられるようにIT技術が指数関数的に成長していくこの世界で、何十万というユーザーの全員に月100MBまでの画像アップロードを許可しているQiitaにとっては、1個あたり2～3ケタKBの画像が有限個残っていたところで大した悪影響は無いのかもしれませんね…。
記事に現役で使っている画像を誤って消してしまうようなリスクも絶対にありますし、そうなったら元に戻せませんから、ユーザーとしては現状、よほど個人情報でも写り込んでいない限りはうかつに触らない方がむしろ無難だと私は感じました。

---

@kato647
「アップロードしたファイル」一覧にある個々の画像が記事に現役で使われているか否か、簡単に判別できるようになりませんか？
究極的には、一覧の各画像に自動で「この画像を使ってる記事or下書きはこれです！」みたいなリンクが付くと最高なんですけど…。

とりあえず画像URLでも見れば手動で照合できるだろうと思ってやってみたのですが、現状はそれさえも意外と大変で、記事の画像クリックで開かれるURLではなくその `<img>` 要素の `data-canonical-src` 属性等にあるURLを参照しないといけませんでした。

**Reply by tsubolabo** [reply]
@Phroneris 

情報をありがとうございました。

　新規投稿記事は、漏らさずにチェックしているつもりでしたが、リリースノートの広報にも、ご紹介の記事にも全く気がつきませんでした。（最新のリリースノートの内容は、いちいちアクセスに行かなくても、黙っていてもトップページに表示されるものと思っていました。）

　早速、覗いてみたところ、消えずに残っている不要画像が沢山あるのが分かりました。しかし、一見しただけでは区別がつかない似たような画像が重複しており、どれが残しておくべき最新の画像なのかが判断できません。下手に削除してしまうのは問題がありそうです。

　間違って、持っているだけで逮捕されるような問題画像をアップしてしまったような事態でもなければ、そのまま放置しておくのが賢明のようです。

**Reply by wataru86** [reply]
@Phroneris

> 「アップロードしたファイル」一覧にある個々の画像が記事に現役で使われているか否か、簡単に判別できるようになりませんか？

現在Qiitaでは、「画像がどの記事に紐づいているか」という情報を保存していません。そのため、提案していただいた機能を実現するには、ユーザーが投稿した記事やコメント、編集リクエスト等すべてのコンテンツの中から各画像URLが含まれているか探す必要があります。

このような理由から、提案していただいた機能をすぐに実装することは困難です。
ご不便をおかけして申し訳ないですが、ご了承いただけますと幸いです🙇

**Reply by Phroneris** [reply]
@wataru86
ご返答ありがとうございます。
実際、すぐの実装は困難だと想像できます。
一方、現状では生きた画像の誤削除のリスクが高いこともご想像いただけるかと存じます。
頻繁に触る機能ではないので開発/ユーザーどちらにとっても優先度は低いと思いますが、ゆくゆくは実現することをのんびり願います。

**Reply by wataru86** [reply]
@Phroneris
ご理解いただきありがとうございます 🙇
提案していただいた機能の検討も含め、引き続き機能の改善に努めてまいりますので、今後ともQiitaをご愛顧いただけますと幸いです ✨
